### DISCUSSION TITLE
Formula for Gini coefficient

### DISCUSSION THREAD

**Discussion by przemyslawj** [petersenpeter/CellExplorer#92] [Q&A] [answered]
Hi Peter,

I wanted to ask about the calculation used in the code, for example, to calculate Gini coefficient of firing rates (https://github.com/petersenpeter/CellExplorer/blob/master/ProcessCellMetrics.m#L1412). The formula seems simplistic, and I am trying to figure out how to derive it from the original definition. I believe this diverges from the formula on Wikipedia for Gini coefficient calculated on sorted values or its unbiased estimator (https://en.wikipedia.org/wiki/Gini_coefficient#Alternative_expressions).

I appreciate your work on the tool and find it extremely helpful in viewing and reasoning about my data!

Kind regards,
Prez

**Comment by petersenpeter** [comment, marked as answer]
Hi Prez

The Gini calculation follows the definition on the Wikipedia page (I use the third definition 1-2B):

Gini = A/(A + B) = 2A = 1 − 2B

where B is the cumulative sorted distribution of X values across the population, which could be the average firing rates across the population (L1408-L1412).

**Comment by przemyslawj** [comment]
Thanks for clarifying; I can understand the reasoning now. A concern with the current implementation might be that the calculated values are somewhat off. For example:
 - perfect equality should give 0, while I get a negative value for x = ones(1,10)
 - perfect inequality should give 1, while I get value = 0.8 for x = [zeros(1,9) 10]
 
 The values get closer to the expected coefficient when the vector is x longer, so this might be less of an issue for real-case scenarios.

**Comment by petersenpeter** [comment]
Hi again Prez

I gave this another go. 

Please see the new calc_gini function: 
 https://github.com/petersenpeter/CellExplorer/commit/f81925bbdf92008e5564a514b79838f64d99390e 


You are correct that equal bins with perfect equality should give 0. I believe I made a mistake in the normalization with the number of samples. I normalized the sum (line 3 below) with the number of cumulated values, n, where I should have normalized by n+1. 

```m
cum_firing1 = cumsum(sort(bined_data(:)));
cum_firing2 = cum_firing1/max(cum_firing1);
gini = 1-2*sum(cum_firing2)./(length(cum_firing2)+1);
```
After this correction the perfect equality gives a gini coefficient = 0. 

Perfect inequality can only be achieved by a very large n, and you will always have a small binning error. 

Let me know what you think.

**Reply by przemyslawj** [reply]
Hi Peter,
The numbers look better now, thank you. Alternatively, you might consider a formula based on the alternative definition of gini on relative mean absolute difference. It would give you exactly 1 for perfect inequality. See formula for G(s) in https://en.wikipedia.org/wiki/Gini_coefficient#Alternative_expressions

`n = length(binned_data);`
`binned_data = sort(binned_data(:));`
`gini = 1 - 2 * n /(n-1) + 2 * sum(binned_data .* (1:n)') / sum(binned_data) / (n-1);`
